### ISSUE TITLE
Allow semantic-release to run if local branch is behind remote

### ISSUE THREAD

**Issue by agaudreault-jive** [semantic-release/semantic-release#1849]
## New feature motivation

When semantic release is added to run automatically on CI for a project with a lot of contributions, the remote branch is quite often outdated and semantic-release will fail to create releases with `The local branch master is behind the remote one, therefore a new version won't be published.`. 

A master build for a small micro-service may take about 10 minutes if you add up the time to build, test, run static code analysis, etc. If at least one PR is merged during that timeframe, the release executed at the end of the build will fail.

The larger the repository, the longer the build may take, and the more contribution it will have, making the situation worse.

Having the build fail during the release step is not convenient for continuous delivery workflow where the master branch is deployed in production on each successful release. Having semantic-release failures because the branch is behind the remote cause the next release to contain more changes and be more impactful.

By merging the changes it can complicate versioning for downstream users as well as make reverting things unnecessarily more complex at the commit level rather than tag level.

## New feature description

Semantic-release should let the user decide if it is problematic to publish a release when the local branch is behind the remote. 
In that case, it is the user responsibility to make sure of the following:

- Plugins that push changes such as `@semantic-release/git` will fail if they are used.
- Concurrent executions are not possible to avoid conflict during releases.

The following changes should be implemented:

- [x] Add a parameter to allow semantic-release to run even if the local branch is behind remote

## New feature implementation

- [x] Validate that the current branch head is not a local commit
- [x] Validate that user has git push permission even if branch is behind remote
- [ ] Add documentation and warning for concurrent/outdated executions
- [ ] Add documentation and warning for plugins that push new commits

Related to https://github.com/semantic-release/semantic-release/issues/1208

**Comment by travi** [comment]
The referenced thread very thoroughly explains why this approach is currently not supported. Adding more configuration options not only makes it easier for our users to end up in unexpected situations, but also adds maintenance and support burden to our team.   

We don't typically recommend applying semver to applications anyway, but worst case, your dependent will succeed once your builds catch up.

**Comment by agaudreault-jive** [comment]
> We don't typically recommend applying semver to applications anyway, but worst case, your dependent will succeed once your builds catch up.

I think the referenced issue already explain that this is not convenient in a continuous delivery workflow where you want your releases to be as small as possible so you do not impact production too much.

`Semantic-release` has awesome features:

- It enforces semantic versioning
- It updates each PR and indicates which version they are part of
- It uses conventional commits to compute the next version
- It creates github tags and release notes

All these features are not `JavaScript` or `npm` centric. That is the main reason we use it. 

However, with the current validation, `semantic-release` is also the only component that can publish your artifacts. I think it would be very beneficial for semantic-release not to have that requirement. It would allow users to add semantic-release on their CI release workflow, with any builder/artifact manager (`npm`, `lerna`, `docker`, `maven`, `gradle`, `msbuild`, `golang`, etc.) like we are doing. For the OSS community and all backend application developers, that would be a huge win.

I think semantic-release should not do assumptions that `@semantic-release/npm` and `@semantic-release/git` are used and perform validations to make sure these plugins won't fail. For me, I think the core value of `semantic-release` is to analyze the git history, generate a version, and perform a release.

So in a non-npm use case where only the plugins `@semantic-release/commit-analyzer`, `@semantic-release/release-notes-generator` and `@semantic-release/github` are used, I think it is an important feature to make sure that `semantic-release` can perform a release for a specified commit, even if it is not the remote's latest.

**Comment by travi** [comment]
> the referenced issue already explain that this is not convenient in a continuous delivery workflow where you want your releases to be as small as possible so you do not impact production too much.

you keep mentioning continuous delivery as if it is a concept that we are unfamiliar with and that it adds a dimension to the equation that justifies a completely different approach that we haven't handled. i am very familiar and use continuous _deployment_ for all applications that i maintain. i've even [given talks](https://ci-vs-cd.travi.org/) about my success with such practices in the past.

one of the biggest benefits of releasing so often with continuous deployment, in my opinion, is that you know that your pipeline will complete, so there is no need to watch it and worry about when it completes. i would suggest that sill holds true in the situation you are experiencing, even if the deployment is delayed a bit.

I recognize that you have a use case where the current implementation of semantic-release may not be ideal for, but please recognize that it has been built to support many workflows, but it a very opinionated way. This is a big part of how the "awesome features" that you reference are possible. While it does support many contexts, it is not intended to be flexible enough to implement any workflow that you may wish.

> I think semantic-release should not do assumptions that @semantic-release/npm and @semantic-release/git are used and perform validations to make sure these plugins won't fail.

I'm unsure where you get the impression that such assumptions are being made. The core bit of the explanation in the thread you reference is that we _cannot_ safely make such assumptions. We have to make some assumptions in core, which include the use of git (not the `@semantic-release/git` plugin), but cannot make assumptions about which plugins are in use. Even checking our known plugins that make git commits cannot ensure that additional plugins are not included that would impact the git history.

> For me, I think the core value of semantic-release is to analyze the git history, generate a version, and perform a release.

personally i recommend, and have had success with, a contract testing tool like [Pact](https://docs.pact.io/) to ensure compatibility with frequent releases. if you find that semver and release notes are a better fit for your teams, this is simply a trade-off that you need to weigh. if semantic-release is valuable to you, the case where you have to wait for another build to complete may simply be a cost you have to endure. if the delayed (but still successful eventually) deployment is a problem, then maybe semantic-release simply isn't the proper tool for this job.

even if there were a safe way for us to detect plugins that would impact git history, it would require a major rearchitecting of how the core of semantic-release works, which at minimum would be a large development effort to justify. simply adding another config flag still adds additional maintenance and support burden and opens up opportunities for mistakes to be made.

**Comment by srcspider** [comment]
> Adding more configuration options not only makes it easier for our users to end up in unexpected situations, but also adds maintenance and support burden to our team.

I don't know how much WORSE then the following that is the current behavior you can get:

1. commit a non breaking change
2. semantic release is now 1.0.0
3. commit a breaking change 
4. pipeline job starts
5. immediately commit a non-breaking change
6. semantic release (silently-)FAILS the pipeline because of the new commit (with success code to console)
7. semantic release does not fail the new non-breaking change since it's latest
8. semantic release is now 1.0.1

There is no way to undo the damage and no visibility that the damage is done since semantic release treats this as "good practice" when it runs in a terminal.

Honestly the complexity of getting semantic version to be "a clean core" (ie. not the npm pusher, package.json dependency it wants to be) is an annoyance in and of itself (and I'm a javascript developer), but that's a minor gripe compared to it completely getting the version wrong.

**Comment by thesalah** [comment]
@srcspider 

> commit a non breaking change
semantic release is now 1.0.0
commit a breaking change
pipeline job starts
immediately commit a non-breaking change
semantic release (silently-)FAILS the pipeline because of the new commit (with success code to console)
semantic release does not fail the new non-breaking change since it's latest
semantic release is now 1.0.1

Not sure how in this scenario the version would be 1.0.1.  Semantic-release would analyze all the commits from the previous release to HEAD to determine the next version. So this would include the breaking change commit as well. So the version would be 2.0.0 right?

**Comment by srcspider** [comment]
That's not a theoretical example. That's a gitlab ci actual case that
happened. Latest version gitlab ci if that makes any difference, but don't
think it does.

**Comment by travi** [comment]
@srcspider can you please post logs? it sounds like you had a partial release, which created a tag for a new release, but did not complete the release process. if that is true, that would be a different failure than the topic that this issue is about. when the branch is behind the remote, the entire release process is skipped, so all commits are analyzed back to the previous release, as @thesalah mentioned.

**Comment by thislooksfun** [comment]
This would also be incredibly useful for dry runs. I'm currently trying to do a dry run to make sure the changelog is correct before pushing some changes to remote (which will trigger an auto-deploy), and I'm getting this error.

I think part of the issue as well is that the message is poorly worded. It states "The local branch <branch> is behind the remote one", regardless of which one is actually behind. This was incredibly confusing at first as my local branch is *ahead* of remote, not behind it.

**Comment by agaudreault-jive** [comment]
@thislooksfun Not sure if it will make it to the master branch and how it works with channels (we dont use those), but I added validation in my fork (https://github.com/semantic-release/semantic-release/pull/1850) so it fails with an error message when the local branch is ahead of the remote and does not silently return.

I also added validation on the tags, so if the remote contains tags that are not present in the local branch, the release will fail. This will prevent to run semantic-release on an old commit that has already been released.

**Comment by mjunior** [comment]
This feature is important for mee too. 
I am trying to use semantic-release in my angular pipeline. My project is shared with 4 teams. Sometimes we have 2 or more PR waiting to merge into master at the same time, for exemplo.

**10:10am:** Team 01 code was merged into master and start the pipeline.
pipeline: checkout, unit test, sonar, semantic-release

**10:12am** Team 02 code was merged into master and start the pipeline.
pipeline: checkout, unit test, sonar, semantic-release

When the first pipeline (team 01) was runnning unit test a new code is merged (team 02), so then, when the semantic release step is exectuted I received the message "the local branch is behind"

It's important to separate them because make sense a git tag 1.1.0 for for Team 01 code, and another git tag 1.2.0 for the team 02 merge.

Is there any way to did that ?

**Comment by agaudreault-jive** [comment]
> This feature is important for mee too.
> I am trying to use semantic-release in my angular pipeline. My project is shared with 4 teams. Sometimes we have 2 or more PR waiting to merge into master at the same time, for exemplo.
> 
> **10:10am:** Team 01 code was merged into master and start the pipeline.
> pipeline: checkout, unit test, sonar, semantic-release
> 
> **10:12am** Team 02 code was merged into master and start the pipeline.
> pipeline: checkout, unit test, sonar, semantic-release
> 
> When the first pipeline (team 01) was runnning unit test a new code is merged (team 02), so then, when the semantic release step is exectuted I received the message "the local branch is behind"
> 
> It's important to separate them because make sense a git tag 1.1.0 for for Team 01 code, and another git tag 1.2.0 for the team 02 merge.
> 
> Is there any way to did that ?

@mjunior Yes, this would fix that. Just make sure to not run your pipeline in parallel (wait before semantic-release step of team 01 is completed before to do checkout of team 02) otherwise the release will fail for team 02 because the branch will miss the 1.1.0 tag

**Comment by jeffrichie** [comment]
Just piling on - this feature is important to me when using semrel in a monorepo with the `semantic-release-monorepo` plugin and multiple GH Actions workflows running at once.

**Comment by kettanaito** [comment]
There must always be an option given to the developer to publish any state of their repository. It's the developer's responsibility to keep track of not rewriting versions and publishing breaking changes accordingly. 

Currently, I can use `semantic-release` at all to manually publish an intermediate version because it refuses to run if the branch is behind its origin. I also can't create a designated release branch manually because it doesn't respect the value of `branches` in the `release.config.js` nethier it respects the `--branches` CLI flag. Basically, the workflow is blocked.

**Comment by Blackclaws** [comment]
We're also using semantic-release to publish applications. Having semantic versioning in place there is helpful for a number of reasons, such as running extra tests on major/minor releases that are costly to run everytime otherwise.

We've had the situation described here happen to us as well where a running build was cancelled by a last minute push request coming in. However for us the situation becomes even more problematic because semantic-release itself can't release all the components in one ci-job. There are multiple components that have specific deployment job that run. Afterwards we run semantic-release to actually do the release for the version. We also run semantic-release before all deployment jobs run to determine the version via a custom plugin that writes the version out.

The issue is that if semantic-release decides that it can't tag this version because its behind then we will already have released a version that semantic-release now doesn't know about. This can break spectacularly as semantic-release will happily assign the same version number if the total change type is the same between this run and the next.

**Comment by danielbayerlein** [comment]
The feature would be really great. Is there any way to complete the PR #1850?

**Comment by jcandan** [comment]
This would be helpful for the use-case described at https://github.com/semantic-release/semantic-release/discussions/1937#discussioncomment-6568960.

**Comment by travi** [comment]
closing as we are not planning to move forward with support of this. please see more context in https://github.com/semantic-release/semantic-release/pull/1850#issuecomment-1666119437

**Comment by ohare93** [comment]
> Just piling on - this feature is important to me when using semrel in a monorepo with the `semantic-release-monorepo` plugin and multiple GH Actions workflows running at once.

Exactly the same situation here. Why can I not simply ignore this error? The only reason local main is behind is because the other GHA run released a completely separate package that happens to live in the same monorepo. We are all big boys that can handle the danger of an `--ignore-git-behind` flag!

**Comment by c0m1c5an5** [comment]
I have been using  [agaudreault's fork](https://github.com/agaudreault/semantic-release/tree/allow-outdated-head) for about two years now, which has been working great so far. Yet it has not been updated for quite some time (as it seems to never have been the goal).

So I wrote a simple [git wrapper](https://gist.github.com/c0m1c5an5/0a1315a135b37b00ba58d07881a1ec6a) that disables the semantic-release HEAD behind remote check, which is hopefully going to provide support for such use case independently of the semantic-release version without requiring constant upkeep of the fork or a patch.

Of course integrating this feature (already implemented by @agaudreault) into semantic-release would be best, but as stated by the maintainers multiple times, this is not going to be added to the project. So the most we could hope for is that they do not start blocking the ways around this.

Installation and usage instructions are in the [gist](https://gist.github.com/c0m1c5an5/0a1315a135b37b00ba58d07881a1ec6a).
The only rule is to not create commits with semantic-release. In that case there is no potential for conflicts to appear.
```sh
#!/usr/bin/env sh

# Copyright (c) 2024 Maksym Kondratenko 
# SPDX-License-Identifier: BSD-3-Clause

set -eu
git_bin="/usr/bin/git"

if ! [ -x "${git_bin}" ]; then
  echo >&2 "ERR: Git binary '${git_bin}' not executable."
  exit 127
fi

if [ "${1:-}" == "ls-remote" ] &&
   [ "${2:-}" == "--heads" ]; then
  git_remote="$("${git_bin}" config --get remote.origin.url)"
  if [ "${3:-}" == "${git_remote}" ]; then
    shift 3
    exec "${git_bin}" show-ref --heads "${@}"
  fi
fi

exec "${git_bin}" "${@}"
```

**Comment by agaudreault** [comment]
I will keep my fork for anyone that wants it. However, i am now using https://github.com/intuit/auto and it seems to allow me to have the same behavior this fork is for. It is the desired behavior of the tool, and is more suitable to use for non js projects.

For my use cases, it has the same or similar capabilities that I needed, while being easier to integrate into a CI.
